In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [18]:
import os

# Carpeta de notebooks
ruta_proyecto = os.path.abspath("..")
print("Ruta proyecto:", ruta_proyecto)

# Listar todo lo que tiene
print("Contenido de proyecto_final:")
print(os.listdir(ruta_proyecto))


Ruta proyecto: c:\Users\ffmma\OneDrive\Desktop\proyecto_final
Contenido de proyecto_final:
['.git', '.gitignore', 'data', 'docs', 'notebooks', 'README.md', 'src']


In [19]:
import os

ruta_data = os.path.join("..", "data")
print("Contenido de data/:")
print(os.listdir(ruta_data))


Contenido de data/:
['data_processed', 'data_raw']


In [20]:
# Carpeta donde están los archivos
ruta_data_raw = os.path.join("..", "data", "data_raw")
ruta_data_raw

'..\\data\\data_raw'

In [21]:
archivo1 = os.path.join(ruta_data_raw, "telecom_clients_us.csv")

df1 = pd.read_csv(archivo1)

df1


user_id tariff_plan  date_start
0     166713           A  2019-08-15
1     166901           A  2019-08-23
2     168527           A  2019-10-29
3     167097           A  2019-09-01
4     168193           A  2019-10-16
..       ...         ...         ...
727   166554           B  2019-08-08
728   166911           B  2019-08-23
729   167012           B  2019-08-28
730   166867           B  2019-08-22
731   166565           B  2019-08-08

[732 rows x 3 columns]

In [22]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      732 non-null    int64 
 1   tariff_plan  732 non-null    object
 2   date_start   732 non-null    object
dtypes: int64(1), object(2)
memory usage: 17.3+ KB


In [23]:
#cambio de tipos de la columna date_start de objetct a datetime 
df1["date_start"]=pd.to_datetime(df1["date_start"], errors="coerce")

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   user_id      732 non-null    int64         
 1   tariff_plan  732 non-null    object        
 2   date_start   732 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 17.3+ KB


In [24]:
#chequo de valores nulos
df1.isnull().sum()
#ninguna columna tiene valores nulos 


user_id        0
tariff_plan    0
date_start     0
dtype: int64

In [25]:
df1.head()

user_id tariff_plan date_start
0   166713           A 2019-08-15
1   166901           A 2019-08-23
2   168527           A 2019-10-29
3   167097           A 2019-09-01
4   168193           A 2019-10-16

In [26]:
archivo2 = os.path.join(ruta_data_raw, "telecom_dataset_us.csv")

df2 = pd.read_csv(archivo2)

df2.head()

user_id                       date direction internal  operator_id  \
0   166377  2019-08-04 00:00:00+03:00        in    False          NaN   
1   166377  2019-08-05 00:00:00+03:00       out     True     880022.0   
2   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
3   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
4   166377  2019-08-05 00:00:00+03:00       out    False     880022.0   

   is_missed_call  calls_count  call_duration  total_call_duration  
0            True            2              0                    4  
1            True            3              0                    5  
2            True            1              0                    1  
3           False            1             10                   18  
4            True            3              0                   25

In [27]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              53902 non-null  int64  
 1   date                 53902 non-null  object 
 2   direction            53902 non-null  object 
 3   internal             53785 non-null  object 
 4   operator_id          45730 non-null  float64
 5   is_missed_call       53902 non-null  bool   
 6   calls_count          53902 non-null  int64  
 7   call_duration        53902 non-null  int64  
 8   total_call_duration  53902 non-null  int64  
dtypes: bool(1), float64(1), int64(4), object(3)
memory usage: 3.3+ MB


In [28]:
df2.sample(10)

user_id                       date direction internal  operator_id  \
22222   167059  2019-10-07 00:00:00+03:00        in    False     911138.0   
48260   168187  2019-11-06 00:00:00+03:00       out    False     937780.0   
15128   166884  2019-10-13 00:00:00+03:00       out    False     895774.0   
5578    166520  2019-11-14 00:00:00+03:00       out    False     892028.0   
25313   167125  2019-10-05 00:00:00+03:00       out    False     902776.0   
27978   167176  2019-11-05 00:00:00+03:00       out     True     907504.0   
49018   168187  2019-11-19 00:00:00+03:00       out    False     937736.0   
33095   167479  2019-11-13 00:00:00+03:00       out    False     915556.0   
39267   167716  2019-11-21 00:00:00+03:00       out    False     922142.0   
4530    166511  2019-09-17 00:00:00+03:00       out    False     891414.0   

       is_missed_call  calls_count  call_duration  total_call_duration  
22222           False            1             38                   60  
48260            True            2              0                   46  
15128           False            1            306                  309  
5578            False           55           1141                 1989  
25313           False           34           2991                 3737  
27978           False           10            114                  155  
49018            True           56              0                 1105  
33095            True           23              0                  854  
39267           False            7            824                  952  
4530             True            8              0                    7

In [29]:
#cambio a datetime de la columna operator_id
df2["date"]= pd.to_datetime(df2["date"],errors = "coerce")
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype                    
---  ------               --------------  -----                    
 0   user_id              53902 non-null  int64                    
 1   date                 53902 non-null  datetime64[ns, UTC+03:00]
 2   direction            53902 non-null  object                   
 3   internal             53785 non-null  object                   
 4   operator_id          45730 non-null  float64                  
 5   is_missed_call       53902 non-null  bool                     
 6   calls_count          53902 non-null  int64                    
 7   call_duration        53902 non-null  int64                    
 8   total_call_duration  53902 non-null  int64                    
dtypes: bool(1), datetime64[ns, UTC+03:00](1), float64(1), int64(4), object(2)
memory usage: 3.3+ MB


In [30]:
#extracción de fecha y horas de la columna date
df2["date"] = df2["date"].dt.date

df2


user_id        date direction internal  operator_id  is_missed_call  \
0       166377  2019-08-04        in    False          NaN            True   
1       166377  2019-08-05       out     True     880022.0            True   
2       166377  2019-08-05       out     True     880020.0            True   
3       166377  2019-08-05       out     True     880020.0           False   
4       166377  2019-08-05       out    False     880022.0            True   
...        ...         ...       ...      ...          ...             ...   
53897   168606  2019-11-10       out     True     957922.0            True   
53898   168606  2019-11-11       out     True     957922.0           False   
53899   168606  2019-11-15       out     True     957922.0           False   
53900   168606  2019-11-15       out     True     957922.0           False   
53901   168606  2019-11-19        in    False          NaN            True   

       calls_count  call_duration  total_call_duration  
0                2              0                    4  
1                3              0                    5  
2                1              0                    1  
3                1             10                   18  
4                3              0                   25  
...            ...            ...                  ...  
53897            1              0                   38  
53898            2            479                  501  
53899            4           3130                 3190  
53900            4           3130                 3190  
53901            2              0                   64  

[53902 rows x 9 columns]

In [31]:
#checar valores nulos
df2.isnull().sum()

user_id                   0
date                      0
direction                 0
internal                117
operator_id            8172
is_missed_call            0
calls_count               0
call_duration             0
total_call_duration       0
dtype: int64

En la columna internal se puede ver que solo hay 117 valores nulos, concidero que es una buena opción mantener estos ya que se puede también analizar si la falta de registro lleva algún patrón. 

La columna presenta un total de 8172 valores nulos, esto lo encuentro un poco preocupante ya que el principal objetivo es valorar el desempeño de los trabajadores. 
Considero que una buena opción para este caso cambiarlos por un valor "0", más adelante podrían ser huzados para poder mostrar la cantidad de llamadas con un operador sin id, para no perder datos y mantener un volumen aunque no se pueda identificar al operador. 

In [34]:
df2["operator_id"] = df2["operator_id"].fillna(0).astype(int)
df2.sample(5)

ValueError: invalid literal for int() with base 10: 'Unknown'